File: src/gmm_visualizer.cpp
// STL
#include <iostream>
#include <vector>
#include <chrono>
#include <random>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/impl/utils.h>
#include <chrono>
#include <functional>
#include <memory>
#include <string>
#include <eigen3/Eigen/Dense>
#include <eigen3/Eigen/Core>
#include <eigen3/Eigen/SVD>
#include <cstdlib>
#include <cmath>
#include <algorithm>
#include <netinet/in.h>
#include <sys/types.h>
#include <stdlib.h>
#include <time.h>
#include <fstream>
#include <stdio.h>
#include <unistd.h>
#include <sys/stat.h>
#include <math.h>
// SFML
// #include <SFML/Graphics.hpp>
// #include <SFML/OpenGL.hpp>
// My includes
#include "gmm_coverage/FortuneAlgorithm.h"
#include "gmm_coverage/Voronoi.h"
// #include "Graphics.h"
// ROS includes
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "geometry_msgs/msg/point.hpp"
#include "sensor_msgs/msg/channel_float32.hpp"
#include "std_msgs/msg/int16.hpp"
#include "std_msgs/msg/bool.hpp"
#include <geometry_msgs/msg/pose.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <visualization_msgs/msg/marker.hpp>
#include <visualization_msgs/msg/marker_array.hpp>
#include "turtlebot3_msgs/msg/gaussian.hpp"
#include "turtlebot3_msgs/msg/gmm.hpp"

#define M_PI   3.14159265358979323846  /*pi*/

using namespace std::chrono_literals;
using std::placeholders::_1;




class Visualizer : public rclcpp::Node
{

public:
    Visualizer() : Node("gmm_visualizer")
    {
        SHAPE = visualization_msgs::msg::Marker::SPHERE;

        // --------------------------------------------------------- GMM ROS publishers and subscribers -------------------------------------------------------
        publisher = this->create_publisher<visualization_msgs::msg::MarkerArray>("/gmm_markers", 1);
        timer_ = this->create_wall_timer(1000ms, std::bind(&Visualizer::show, this));
        gmmSub_ = this->create_subscription<turtlebot3_msgs::msg::GMM>("/gaussian_mixture_model", 1, std::bind(&Visualizer::gmm_callback, this, _1));

    }
    ~Visualizer()
    {
        std::cout<<"DESTROYER HAS BEEN CALLED"<<std::endl;
    }

    void gmm_callback(const turtlebot3_msgs::msg::GMM::SharedPtr msg);
    void GroundColorMix(double* color, double x, double min, double max);
    void show();



private:

    uint32_t SHAPE;
    std::vector<std::vector<float>> MEANs;
    std::vector<std::vector<std::vector<float>>> VARs;
    std::vector<float> weights;

    //------------------------- Publishers and subscribers ------------------------------
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr publisher;
    rclcpp::Subscription<turtlebot3_msgs::msg::GMM>::SharedPtr gmmSub_;
    turtlebot3_msgs::msg::GMM gmm_msg;
    visualization_msgs::msg::MarkerArray markers_msg;
    std::vector<visualization_msgs::msg::Marker> markers_array;
    //-----------------------------------------------------------------------------------

};

void Visualizer::gmm_callback(const turtlebot3_msgs::msg::GMM::SharedPtr msg)
{
    this->gmm_msg.gaussians = msg->gaussians;
    this->gmm_msg.weights = msg->weights;
}

// Returns RGB color in colormap from green to red for a given value x
void Visualizer::GroundColorMix(double* color, double x, double min = 0.0, double max = 1.0)
{
   /*  */
    double posSlope = (max-min)/60;
    double negSlope = (min-max)/60;

    if( x < 60 )
    {
        color[0] = max;
        color[1] = posSlope*x+min;
        color[2] = min;
        return;
    }
    else if ( x < 120 )
    {
        color[0] = negSlope*x+2*max+min;
        color[1] = max;
        color[2] = min;
        return;
    }
    else if ( x < 180  )
    {
        color[0] = min;
        color[1] = max;
        color[2] = posSlope*x-2*max+min;
        return;
    }
    else if ( x < 240  )
    {
        color[0] = min;
        color[1] = negSlope*x+4*max+min;
        color[2] = max;
        return;
    }
    else if ( x < 300  )
    {
        color[0] = posSlope*x-4*max+min;
        color[1] = min;
        color[2] = max;
        return;
    }
    else
    {
        color[0] = max;
        color[1] = min;
        color[2] = negSlope*x+6*max;
        return;
    }
}

void Visualizer::show()
{

    for (int i = 0; i < this->gmm_msg.gaussians.size(); i++)
    {
        // -------------- Draw ellipse from covariance matrix ----------------
        // Elements of the covariance matrix
        double a = this->gmm_msg.gaussians[i].covariance[0];
        double b = this->gmm_msg.gaussians[i].covariance[1];
        double c = this->gmm_msg.gaussians[i].covariance[3];

        // Ellipse's parameters
        double lambda1 = (a+c)/2 + sqrt( pow((a-c)/2,2) + pow(b,2) );
        double lambda2 = (a+c)/2 - sqrt( pow((a-c)/2,2) + pow(b,2) );
        double r1 = sqrt(lambda1);
        double r2 = sqrt(lambda2);
        double theta = 0.0;
        if (b == 0.0 && a >= c)
            theta = 0.0;
        else if (b == 0.0 && a < c)
            theta = M_PI/2;
        else
            theta = atan2((lambda1-a), b);
        visualization_msgs::msg::Marker m;
        m.id = i;
        m.header.frame_id = "odom";
        m.type = SHAPE;
        m.action = visualization_msgs::msg::Marker::ADD;
        m.pose.position.x = this->gmm_msg.gaussians[i].mean_point.x;
        m.pose.position.y = this->gmm_msg.gaussians[i].mean_point.y;
        m.pose.position.z = this->gmm_msg.gaussians[i].mean_point.z;
        if (this->gmm_msg.gaussians[i].covariance.size() == 9)
        {
            // 3D case
            m.scale.x = this->gmm_msg.gaussians[i].covariance[0];
            m.scale.y = this->gmm_msg.gaussians[i].covariance[4];
            m.scale.z = this->gmm_msg.gaussians[i].covariance[8];
        } else
        {   
            // 2D case
            // m.scale.x = this->gmm_msg.gaussians[i].covariance[0];
            // m.scale.y = this->gmm_msg.gaussians[i].covariance[3];
            m.scale.x = r1;
            m.scale.y = r2;
            m.scale.z = 0.01;
            tf2::Quaternion q;
            q.setRPY(0, 0, theta);
            m.pose.orientation.x = q.x();
            m.pose.orientation.y = q.y();
            m.pose.orientation.z = q.z();
            m.pose.orientation.w = q.w();
        }
        // m.color.r = 0.0f;
        // m.color.g = 1.0f;
        // m.color.b = 0.0f;
        // m.color.a = this->gmm_msg.weights[i];           // alpha is given by its weight
        double color[3];
        GroundColorMix(color, 360.0 * this->gmm_msg.weights[i]);
        m.color.r = color[0];
        m.color.g = color[1];
        m.color.b = color[2];
        m.color.a = 1.0;
        this->markers_msg.markers.push_back(m);
    }
    this->publisher->publish(this->markers_msg);
}



int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Visualizer>());
    rclcpp::shutdown();
    return 0;
}

File: scripts/gmm_visualizer.py
#!/usr/bin/env python3


import rospy
import sys
import math
from geometry_msgs.msg import Point
from gmm_msgs.msg import GMM, Gaussian
from visualization_msgs.msg import Marker, MarkerArray


class GMM_Vis():
    def __init__(self):
        rospy.init_node("gmm_visualizer")
        
        pub = rospy.Publisher("/gmm_visualizer", MarkerArray, queue_size=1)       
        sub = rospy.Subscriber("/gaussian_mixture_model", GMM, self.callback)

        self.pub = pub
        self.sub = sub
        
        self.gmm_msg = GMM()
        self.marker_msg = MarkerArray()

    def callback(self, msg):
        self.gmm_msg.gaussians = msg.gaussians
        self.gmm_msg.weights = msg.weights

    def GroundColorMix(self, x, min = 0.0, max = 1.0):
        posSlope = (max-min)/60
        negSlope = (min-max)/60
        color = [0.0, 0.0, 0.0]

        if x < 60:
            color[0] = max
            color[1] = posSlope + min
            color[2] = min
            return color
        elif x < 120:
            color[0] = negSlope * x + 2 * max + min
            color[1] = max
            color[2] = min
            return color
        elif x < 180:
            color[0] = min
            color[1] = max
            color[2] = posSlope * x - 2 * max + min
            return color
        elif x < 240:
            color[0] = min
            color[1] = negSlope * x + 4 * max + min
            color[2] = max
            return color
        elif x < 300:
            color[0] = posSlope * x - 4 * max + min
            color[1] = min
            color[2] = max
            return color
        else:
            color[0] = max
            color[1] = min
            color[2] = negSlope * x + 6 * max
            return color
        
    def euler_to_quat(self, roll, pitch, yaw):
        cy = math.cos(yaw * 0.5)
        sy = math.sin(yaw * 0.5)
        cp = math.cos(pitch * 0.5)
        sp = math.sin(pitch * 0.5)
        cr = math.cos(roll * 0.5)
        sr = math.sin(roll * 0.5)

        w = cy * cp * cr + sy * sp * sr
        x = cy * cp * sr - sy * sp * cr
        y = sy * cp * sr + cy * sp * cr
        z = sy * cp * cr - cy * sp * sr

        return [w, x, y, z]

        

    
    def show(self):
        for i in range(len(self.gmm_msg.weights)):
            # Draw ellipse from covariance matrix
            a = self.gmm_msg.gaussians[i].covariance[0]
            b = self.gmm_msg.gaussians[i].covariance[1]
            c = self.gmm_msg.gaussians[i].covariance[3]

            # Ellipses parameters
            lambda1 = (a+c)/2 + ((a-c)**2 + 4*b**2)**0.5/2
            lambda2 = (a+c)/2 - ((a-c)**2 + 4*b**2)**0.5/2
            r1 = lambda1**0.5
            r2 = lambda2**0.5
            theta = 0.0
            if b == 0.0 and a >= c:
                theta = 0.0
            elif b == 0.0 and a < c:
                theta = 0.5 * math.pi
            else:
                theta = math.atan2(lambda1 - a, b)

            m = Marker()
            m.id = i
            m.header.frame_id = "map"
            m.type = m.SPHERE
            m.action = m.ADD
            m.pose.position.x = self.gmm_msg.gaussians[i].mean_point.x
            m.pose.position.y = self.gmm_msg.gaussians[i].mean_point.y
            m.pose.position.z = 0.0

            m.scale.x = r1
            m.scale.y = r2
            m.scale.z = 0.1

            [qw, qx, qy, qz] = self.euler_to_quat(0.0, 0.0, theta)
            m.pose.orientation.w = qw
            m.pose.orientation.x = qx
            m.pose.orientation.y = qy
            m.pose.orientation.z = qz

            color = self.GroundColorMix(360.0 * self.gmm_msg.weights[i])
            m.color.r = color[0]
            m.color.g = color[1]
            m.color.b = color[2]
            m.color.a = 1.0

            self.marker_msg.markers.append(m)
            self.pub.publish(self.marker_msg)







def main():
    viz = GMM_Vis()
    
    while not rospy.is_shutdown():
        viz.show()
        rospy.sleep(1)
    # rospy.spin()


if __name__ == '__main__':
    main()
    


